### PR TITLE
Add Air Sweep

### PR DESCRIPTION
**Pull request by itanasi**
Solves #6553
Units with the Can Air Sweep Unique can now Air Sweep a tile and draw Interceptions
* Dogfighting Promotions added and supported
* Toggle unit into Air Sweep Mode via new Action
* Takes an attack, and Interceptor uses up an attack
* Can attack any visible tile in range
* New airSweep battle function
  * Land and Sea AA deal no damage and neither combatant gains XP
  * Air Interception fight as if Melee and both gain 5 XP
  * A lot of new Notifications
  * Interceptors can't miss, meaning you draw Air Units into battle and deal damage
* For now, AI doesn't execute Air Sweeps. Just waits to Intercept instead. Coordination with Bombers is tough

image
image

### CODE REVIEW

**Comment by itanasi** [comment]
Discussion point. Should Interceptors be able to miss when getting drawn out by an Air Sweep.
On one hand, if they miss, no damage (and no XP) for either party. But still use up and attack.
However, drawing out an Interception means taking damage as the attacker, but also means the Interceptor takes damage as well.
Civ5 Interceptors always come out and hit.

**Comment by itanasi** [comment]
That's intentional. Non-Air Interception does no damage, just burns an attack

**Comment by xlenstra** [comment]
> That's intentional. Non-Air Interception does no damage, just burns an attack

This differs from Civ V behaviour

<details> <summary> Picutre: </summary>

image
This is a planned ranged attack by a jet fighter on a marine next to a SAM. As you can see, the damage is marked as increased due to a known interception. Executing the attack doesn't give ~37 damage as predicted, but destroys the fighter, so the interception alone dealt at least ~63 damage.
</details>

File: core/src/com/unciv/logic/battle/Battle.kt [hunk 1]
```diff
@@ -810,6 +810,143 @@ object Battle {
...
+    // Land Units deal 0 damage, and no XP for either party
+    // Air Interceptors do Air Combat as if Melee but using Ranged Strength. 5XP to both
+    // But does not use the Interception mechanic bonuses/promotions
+    // Counts as an Attack for both units
+    // Will always draw out an Interceptor's attack (they cannot miss)
+    // This means the combat against Air Units will execute and always deal damage
+    fun airSweep(attacker: MapUnitCombatant, attackedTile: TileInfo) {
+        // Air Sweep counts as an attack, even if nothing else happens
+        attacker.unit.attacksThisTurn++
+        // copied and modified from reduceAttackerMovementPointsAndAttacks()
```
[the remark this answers is not in the document]
Comment (itanasi) [reply core/src/com/unciv/logic/battle/Battle.kt:823]: The extra notifications and sheer amount of stuff not used would mean If statement hell to me imo

File: core/src/com/unciv/logic/battle/Battle.kt [hunk 2]
```diff
@@ -810,6 +810,143 @@ object Battle {
...
+                        interceptingCiv.addNotification(
+                            interceptorText, locationsAttackerUnk,
+                            interceptorName, NotificationIcon.War, NotificationIcon.Question
+                        )
+                    }
+                } else {
+                    val attackerText =
+                            "Our [$attackerName] was attacked by an intercepting [$interceptorName]"
+                    val interceptorText =
+                            "Our [$interceptorName] intercepted and attacked an enemy [$attackerName]"
```
[the remark this answers is not in the document]
Comment (itanasi) [reply core/src/com/unciv/logic/battle/Battle.kt]: Done

File: core/src/com/unciv/logic/battle/Battle.kt [hunk 3]
```diff
@@ -810,6 +810,143 @@ object Battle {
...
+
+        // Handle which Civ Intercepts
+        for (interceptingCiv in UncivGame.Current.gameInfo!!.civilizations) {
+            // only try to intercept if at war
+            if (!attacker.getCivInfo().isAtWarWith(interceptingCiv))
+                continue
+            // Pick highest chance interceptor
+            for (interceptor in interceptingCiv.getCivUnits()
+                .filter { it.canIntercept(attackedTile) }
+                .sortedByDescending { it.interceptChance() }) {
```
[the remark this answers is not in the document]
Comment (itanasi) [reply core/src/com/unciv/logic/battle/Battle.kt]: And done

File: android/assets/jsons/Civ V - Gods & Kings/UnitPromotions.json [hunk 4]
```diff
@@ -358,29 +358,26 @@
 		"uniques": ["[+34]% Damage when intercepting"],
 		"unitTypes": ["Fighter"]
 	},
-	/*
 	{
 		"name": "Dogfighting I",
-		"uniques": ["Bonus when performing air sweep [33]%"], // todo
+		"uniques": ["[+33]% Strength bonus when performing air sweep"],
```
Comment (xlenstra) [inline android/assets/jsons/Civ V - Gods & Kings/UnitPromotions.json:363]: We generally try not to use the words 'bonus' in uniques, as "-33% Strengh bonus" sounds a bit weird. I'd write this as "[+33]% Strength when performing an air sweep"

File: android/assets/jsons/Civ V - Gods & Kings/Units.json [hunk 5]
```diff
@@ -1194,7 +1194,7 @@
 		"requiredResource": "Oil",
 		"upgradesTo": "Fighter",
 		"obsoleteTech": "Radar",
-		"uniques": ["[50]% chance to intercept air attacks", "[+150]% Strength <vs [Bomber] units>", "[+150]% Strength <vs [Helicopter] units>"],
+		"uniques": ["Can perform Air Sweep", "[50]% chance to intercept air attacks", "[+150]% Strength <vs [Bomber] units>", "[+150]% Strength <vs [Helicopter] units>"],
```
Comment (xlenstra) [inline android/assets/jsons/Civ V - Gods & Kings/Units.json:1197]: Here Air Sweep is capitalized, but in the previous unique it is not?

File: android/assets/jsons/Civ V - Vanilla/Units.json [hunk 6]
```diff
@@ -1241,7 +1241,7 @@
 		"cost": 425,
 		"requiredTech": "Lasers",
 		"requiredResource": "Aluminum",
-		"uniques": ["[100]% chance to intercept air attacks", "[+150]% Strength <vs [Bomber] units>", "[+150]% Strength <vs [Helicopter] units>"],
+		"uniques": ["Can perform Air Sweep", "[100]% chance to intercept air attacks", "[+150]% Strength <vs [Bomber] units>", "[+150]% Strength <vs [Helicopter] units>"],
```
Comment (xlenstra) [inline android/assets/jsons/Civ V - Vanilla/Units.json:1244]: I'd rather put this unique on the unit type than the unit btw

File: core/src/com/unciv/logic/battle/Battle.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (itanasi) [reply core/src/com/unciv/logic/battle/Battle.kt]: Can also sort by strength

File: android/assets/jsons/Civ V - Gods & Kings/UnitPromotions.json (hunk 4 above)
Comment (itanasi) [reply android/assets/jsons/Civ V - Gods & Kings/UnitPromotions.json]: done

File: android/assets/jsons/Civ V - Gods & Kings/Units.json (hunk 5 above)
Comment (itanasi) [reply android/assets/jsons/Civ V - Gods & Kings/Units.json]: Capitalized all

File: android/assets/jsons/Civ V - Vanilla/Units.json (hunk 6 above)
Comment (itanasi) [reply android/assets/jsons/Civ V - Vanilla/Units.json]: Done

File: core/src/com/unciv/logic/battle/Battle.kt [hunk 7]
```diff
@@ -810,6 +816,151 @@ object Battle {
...
+            .sortedByDescending { it.interceptChance() }) {
+            // No chance of Interceptor to miss (unlike regular Interception). Always want to deal damage
+            val interceptingCiv = interceptor.civInfo
+            val interceptorName = interceptor.name
+            // pairs of LocationAction for Notification
+            val locations = LocationAction(
+                interceptor.currentTile.position,
+                attacker.unit.currentTile.position
+            )
+            val locationsAttackerUnk =
```
Comment (xlenstra) [inline core/src/com/unciv/logic/battle/Battle.kt:872]: What does 'unk' stand for?

File: core/src/com/unciv/logic/battle/Battle.kt [hunk 8]
```diff
@@ -306,7 +306,13 @@ object Battle {
         return true
     }
 
-    private fun takeDamage(attacker: ICombatant, defender: ICombatant): Pair<Int, Int> {
+    private class DamageDealt(attackerDealt: Int, defenderDealt: Int)
```
Comment (xlenstra) [inline core/src/com/unciv/logic/battle/Battle.kt:309]: I'd just make this is data class, that saves declaring the additional variables:

```kotlin
private data class DamageDealt(val attackerDealt: Int, val defenderDealt: Int) {}
```

**Review by xlenstra — APPROVED** [review]
Barring my last two comments, looks good! Haven't tested it tough. Feel free to merge once you've addressed these comments

**Pull request merged**